### ISSUE TITLE
Provisioning-API PUT not working

### ISSUE THREAD

**Issue by LiegenderHorst** [owncloud/core#14283]
Hi,

I just installed Owncloud 8.0 and was happy to see that they now provide their provisioning API with the community edition.
Everything is really straight forward and works like a charm (well yeah...the post parameter "user" does not exist - "userid" would be the correct one. documentation is wrong on that).

After checking all functions I have a problem with getting an "edit user" call to work. 
On my local owncloud installation (apache2, no ssl) "edit user" works fine but in my dev environment where I use nginx with ssl and a proxy it always results in an 997 error without any message (message: null).

my example api call: 
```curl -X PUT --user <username>:<password> https://cloud.my-dev.de/ocs/v1.php/cloud/users/foobar -d key="display" -d value="foobarbaz" --insecure```

Any other API call (GET, POST, DELETE) works fine - just "edituser" gives me a 997 without any error message.

In the owncloud log I can see the following php error:
```Undefined index: key at /html/apps/provisioning_api/lib/users.php#151```

That would be an invalid index in my PUT-Body according to the API Users class (owncloud/apps/provisioning_api/lib/users.php)...

So I did some debugging in the owncloud sources.

Local call:
```
MyMachine git:(owncloud) ✗ curl -X PUT --user user:password http://localhost/owncloud/ocs/v1.php/cloud/users/foobar -d key="display" -d value="foobarbaz"

$request:                   
Array
(
    [userid] => foobar
    [_route] => put_cloud_users__userid_
    [_put] => Array
        (
            [key] => display
            [value] => foobarbaz
        )

)
Array
(
    [0] => display
    [1] => quota
    [2] => password
    [3] => email
)

$response
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>ok</status>
  <statuscode>100</statuscode>
  <message/>
 </meta>
 <data/>
</ocs>
```

Dev-Call:
```
➜  MyMachine git:(owncloud) ✗ curl -X PUT --user user:password https://owncloud-dev.de/ocs/v1.php/cloud/users/foobar -d key="display" -d value="foobarbaz" --insecure

$request
Array
(
    [userid] => foobar
    [_route] => put_cloud_users__userid_
    [_put] => Array
        (
        )

)
Array
(
    [0] => display
    [1] => quota
    [2] => password
    [3] => email
)

$response
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>failure</status>
  <statuscode>997</statuscode>
  <message/>
 </meta>
 <data/>
</ocs>
```

In nginx-log I can see:
```
[17/Feb/2015:11:35:59 +0100] "PUT /ocs/v1.php/cloud/users/foobar HTTP/1.1" 401 283 "-" "curl/7.35.0" "-"
```

If I verbose my curl request, I get:
```
* Connected to somehost (xxx.xxx.xxx.xxx) port 443 (#0)
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
* SSLv3, TLS handshake, Server hello (2):
* SSLv3, TLS handshake, CERT (11):
* SSLv3, TLS handshake, Server key exchange (12):
* SSLv3, TLS handshake, Server finished (14):
* SSLv3, TLS handshake, Client key exchange (16):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv3, TLS handshake, Finished (20):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv3, TLS handshake, Finished (20):
* SSL connection using ECDHE-RSA-AES128-GCM-SHA256
* Server certificate:
* 	 bla...
* 	 SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* Server auth using Basic with user 'username'
> PUT /ocs/v1.php/cloud/users/foobar HTTP/1.1
> Authorization: Basic somehashXYZ
> User-Agent: curl/7.35.0
> Host: my.host
> Accept: */*
> Content-Length: 27
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 27 out of 27 bytes
< HTTP/1.1 401 Unauthorized
* Server nginx is not blacklisted
< Server: nginx
< Date: Tue, 17 Feb 2015 10:35:59 GMT
< Content-Type: text/xml; charset=UTF-8
< Transfer-Encoding: chunked
< Connection: keep-alive
< Expires: Thu, 19 Nov 1981 08:52:00 GMT
< Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
< Pragma: no-cache
< Strict-Transport-Security: max-age=0
< X-XSS-Protection: 1; mode=block
< X-Content-Type-Options: nosniff
< X-Frame-Options: Sameorigin
< Content-Security-Policy: default-src 'self'; script-src 'self' 'unsafe-eval'; style-src 'self' 'unsafe-inline'; frame-src *; img-src *; font-src 'self' data:; media-src *; connect-src *
< X-Robots-Tag: none
< Set-Cookie: oc_username=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0
< Set-Cookie: oc_token=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0
< Set-Cookie: oc_remember_login=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0
< Set-Cookie: oc_username=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/
< Set-Cookie: oc_token=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/
< Set-Cookie: oc_remember_login=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/
< Set-Cookie: somecookie=foobarxyz; path=/; secure; HttpOnly
* Authentication problem. Ignoring this.
< WWW-Authenticate: Basic realm="Authorisation Required"
< 
```

So nginx seems to kill my request body somehow...but I have no idea how to fix that. 
Any ideas on that are very much appreciated. 

Kind regards

**Comment by karlitschek** [comment]
@tomneedham can you have look?

**Comment by tomneedham** [comment]
@LiegenderHorst Which version of nginx are you using?

**Comment by LiegenderHorst** [comment]
@tomneedham 
I use nginx version: nginx/1.6.0 built by gcc 4.7.2 (Debian 4.7.2-5) with PHP 5.5.21-1~dotdeb.1 (fpm-fcgi) (php5-fpm) and owncloud runs in a chroot.
I also use a proxy which is set in env as http and https proxy. Network communication runs fine though...

My nginx config is very much like the owncloud example config with some more security parameters like server-tokens off and some more ssl_ciphers.
I stored them in a separate file "owncloud.conf" and load it in nginx http context.

```
http {
    server_names_hash_bucket_size 64;
    include       /etc/nginx/mime.types;
    default_type  application/octet-stream;

    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

    access_log  /var/log/nginx/access.log  main;

    sendfile        on;

    keepalive_timeout  65;

    server_tokens off;

    include /etc/nginx/conf.d/*.conf;
}

```

I have not set any basic_auth options.

**Comment by LiegenderHorst** [comment]
Another strange thing is that if I call a PUT method, owncloud log says "app":"PHP" instead of "app":"ocs_api". 

```
{"reqId":"c7724b50f2b2424a4337106aa16617bc","remoteAddr":"xxx.xxx.xxx.xxx","app":"ocs_api","message":"Successful addUser call with userid: foobar","level":1,"time":"2015-02-19T09:21:25+00:00","method":"POST","url":"\/ocs\/v1.php\/cloud\/users?format=json"}
{"reqId":"3dbf246a0dc89f4811ca07b73c120be9","remoteAddr":"xxx.xxx.xxx.xxx","app":"PHP","message":"Undefined index: key at \/html\/apps\/provisioning_api\/lib\/users.php#152","level":3,"time":"2015-02-19T09:21:26+00:00","method":"PUT","url":"\/ocs\/v1.php\/cloud\/users\/foobar?format=json"}
```

**Comment by LiegenderHorst** [comment]
Ok - I found the problem...

The whole issue only occurs if the calendar app is activated - even on a local installation without any security options.


**How to reproduce:**
1.  Fresh owncloud 8.0 installation or upgrade
2.  Activate Calendar App
3.  Do a PUT request through provisioning API on an existing user (example below for changing display name to "foobar"):
```
curl -v -X PUT --user user:password http://<your-host>/owncloud/ocs/v1.php/cloud/users/<any_existing_user> -d key="display" -d value="foobar"
```


**Result:**
1.  You will receive a response 997 with an empty message
2.  Owncloud log will show something like: 
```
{"reqId":"2802bb385b073dba6de4abd76600e29f","remoteAddr":"xxx.xxx.xxx.xxx","app":"PHP","message":"Invalid argument supplied for foreach() at \/html\/themes\/mw\/core\/templates\/layout.guest.php#26","level":3,"time":"2015-02-19T10:52:03+00:00","method":"GET","url":"\/"}
```
The relevant part here is the ```"app":"PHP"``` which means that this request is not handled by the provisioning API. Correct API calls always have a part ```"app":"ocs_api"``` 
3.  your curl request will have an empty body
4.  the result of your curl request will be 401 and last two line will be something like this:
```
* Authentication problem. Ignoring this.
< WWW-Authenticate: Basic realm="Authorisation Required"
```

**Comment by LiegenderHorst** [comment]
Any news on this? 
Should I repost this on the calendar issue tracker?

**Comment by tomneedham** [comment]
@georgehrke (For calendar) and @DeepDiver1975 who might have a hint on why the routing is being affected here? I haven't had time to look into this yet.

**Comment by LiegenderHorst** [comment]
Just to clarify myself: the provisioning API works fine as its (in-app) routing does. 

Maybe this could be a hint where to look:
```
{"reqId":"05a272d37762984dbc029259ff42a1bd","remoteAddr":"xxx.xxx.xxx.xxx","app":"OCP\\Share","message":"Sharing backend OC_Share_Backend_Calendar not registered, OC_Share_Backend_Calendar is already registered for calendar","level":2,"time":"2015-02-19T12:26:34+00:00","method":"GET","url":"\/index.php\/settings\/ajax\/navigationdetect.php?app=calendar"}
{"reqId":"05a272d37762984dbc029259ff42a1bd","remoteAddr":"xxx.xxx.xxx.xxx","app":"OCP\\Share","message":"Sharing backend OC_Share_Backend_Event not registered, OC_Share_Backend_Event is already registered for event","level":2,"time":"2015-02-19T12:26:34+00:00","method":"GET","url":"\/index.php\/settings\/ajax\/navigationdetect.php?app=calendar"}
```

**Comment by DeepDiver1975** [comment]
marking this as duplicate #14541

**Comment by DeepDiver1975** [comment]
@LukasReschke looks like the request php://input issue is on stable8 as well - maybe calendar is instantiating the request object as well and ready up all data from php://input

@Raydiation you might want to have a look as well on how we can ensure php://input is really only read once - THX a ton

**Comment by LukasReschke** [comment]
https://github.com/owncloud/calendar/blob/0c8d09d384e38cd72c5ee196dbba59ef09a12d92/appinfo/app.php#L41 initializes the Request Class which does https://github.com/owncloud/core/blob/3356abe5c705b37bb1c41529b3c836170c4e219e/lib/private/appframework/http/request.php#L95

**Comment by LukasReschke** [comment]
As far I can tell https://github.com/owncloud/core/pull/14574 should "fix" that as well

**Comment by Raydiation** [comment]
tricky bug, @DeepDiver1975 good job for figuring this out :)

**Comment by DeepDiver1975** [comment]
@Raydiation can we kind of delay reading from php://input in the request implementation to no happen in ctor? This should require less inversive changes - which would be desired for backport :wink:

**Comment by LukasReschke** [comment]
Should be possible as well. Let me look into that.

**Comment by LukasReschke** [comment]
… mhm … we also have some other issues of this category … *sigh*

**Comment by LukasReschke** [comment]
The hacky ownCloud 8 approach: https://github.com/owncloud/core/pull/14587

**Comment by DeepDiver1975** [comment]
This one is closed from my perspective with #14587 and related PRs
